File: leet/1010_pairs_divisible_by_60/1010_pairs_divisible_by_60.js
// LeetCode 1010. Pairs of Songs With Total Durations Divisible by 60
// https://leetcode.com/problems/pairs-of-songs-with-total-durations-divisible-by-60/
// O(n) time, O(1) space
var numPairsDivisibleBy60 = function(time) {
    const modToCount = Array(60).fill(0);
    let pairs = 0;
    for (let t of time) {
        t %= 60;
        modToCount[t]++;
        if (t !== 0 && t !== 30)
            pairs += modToCount[60 - t];
    }
    pairs += modToCount[0] * (modToCount[0] - 1) / 2;
    pairs += modToCount[30] * (modToCount[30] - 1) / 2;
    return pairs;
};

File: leet/1010_pairs_divisible_by_60/1010_pairs_divisible_by_60.cpp
// LeetCode 1010. Pairs of Songs With Total Durations Divisible by 60
// https://leetcode.com/problems/pairs-of-songs-with-total-durations-divisible-by-60/
// O(n) time, O(1) space
class Solution {
public:
    int numPairsDivisibleBy60(vector<int>& time) {
        int count[60] = {0};
        for (int t : time)
            count[t % 60]++;
        int pairs = count[30] * (count[30] - 1) / 2;
        pairs += count[0] * (count[0] - 1) / 2;
        for (int t = 1; t < 30; t++)
            pairs += count[t] * count[60 - t];
        return pairs;
    }
};
